From 8b3b6531f3b6e3459e24baf2719bfaa806eaf0d4 Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Mon, 22 Nov 2021 16:44:51 +1100
Subject: [PATCH 05/15] msdasql: Implement ICommandPrepare Prepare

The handle should be make NULL when it's passed to EXECUTE.
---
 dlls/msdasql/session.c | 23 +++++++++++++++++++++++
 1 file changed, 23 insertions(+)

diff --git a/dlls/msdasql/session.c b/dlls/msdasql/session.c
index 98b2b4e6c5d..281d99a0bd8 100644
--- a/dlls/msdasql/session.c
+++ b/dlls/msdasql/session.c
@@ -402,6 +402,7 @@ struct command
     LONG refs;
     WCHAR *query;
     IUnknown *session;
+    SQLHSTMT hstmt;
 
     struct msdasql_prop properties[80];
 };
@@ -523,6 +524,10 @@ static ULONG WINAPI command_Release(ICommandText *iface)
         TRACE( "destroying %p\n", command );
         if (command->session)
             IUnknown_Release(command->session);
+
+        if (command->hstmt)
+            SQLFreeHandle(SQL_HANDLE_STMT, command->hstmt);
+
         heap_free( command->query );
         heap_free( command );
     }
@@ -1304,7 +1309,24 @@ static ULONG WINAPI commandprepare_Release(ICommandPrepare *iface)
 static HRESULT WINAPI commandprepare_Prepare(ICommandPrepare *iface, ULONG runs)
 {
     struct command *command = impl_from_ICommandPrepare( iface );
+    RETCODE ret;
+
     TRACE("%p, %u\n", command, runs);
+
+    if (!command->query)
+        return DB_E_NOCOMMAND;
+
+    if (command->hstmt)
+        SQLFreeHandle(SQL_HANDLE_STMT, command->hstmt);
+
+    SQLAllocHandle(SQL_HANDLE_STMT, command->hdbc, &command->hstmt);
+
+    ret = SQLPrepareW(command->hstmt, command->query, lstrlenW(command->query));
+    if (ret != SQL_SUCCESS)
+    {
+        dump_sql_diag_records(SQL_HANDLE_STMT, command->hstmt);
+        return E_FAIL;
+    }
     return S_OK;
 }
 
@@ -1400,6 +1422,7 @@ static HRESULT WINAPI createcommand_CreateCommand(IDBCreateCommand *iface, IUnkn
     command->ICommandWithParameters_iface.lpVtbl = &command_with_params_vtbl;
     command->refs = 1;
     command->query = NULL;
+    command->hstmt = NULL;
 
     memcpy(command->properties, msdasql_init_props, sizeof(msdasql_init_props));
 
-- 
2.33.0

